### DISCUSSION TITLE
Vanishing modes: lower log_Z_uncert and resume incrementally

### DISCUSSION THREAD

**Discussion by Joshuaalbert** [Joshuaalbert/jaxns#274] [General]
On a difficult multimodal problem, a narrow or low-mass mode can disappear
from a nested-sampling run. In JAXNS v3, the simplest response is often to ask
for a lower expected log-evidence uncertainty, `log_Z_uncert`.

A tighter goal causes more allocation rounds and increases the number of
lineages. Those additional lineages give the constrained sampler more chances
to rediscover and retain an isolated mode. There is no universal target: start
with a reasonable value, inspect the result, and lower it incrementally if a
scientifically plausible mode is absent or unstable across seeds.

## Incremental pattern

Keep the custom checkpoint separate from the ordinary compiled depth rule so
the returned state remains resumable:

```python
import jax
import jax.numpy as jnp

from jaxns.core import NestedSampler
from jaxns.termination_condition import TerminationCondition


def evidence_goal(target_log_Z_uncert):
    def goal(state):
        # Always complete at least one allocation round.
        if int(state.goal_loop_iter) == 0:
            return False
        register = state.compute_termination_register()
        return float(register.log_Z_uncert) <= target_log_Z_uncert

    return goal


# The sampler-level condition contains no ordinary convergence goal, so a
# custom uncertainty checkpoint is a continuation point rather than terminal.
sampler = NestedSampler(
    model=model,
    unlimited_samples=True,
    termination_condition=TerminationCondition(),
)

# Keep the usual remaining-evidence rule inside each compiled depth epoch.
depth_condition = TerminationCondition(dlogZ=jnp.log1p(1e-3))

state = sampler.run_until_goal(
    evidence_goal(0.10),
    depth_cond=depth_condition,
    key=jax.random.PRNGKey(42),
)

# Inspect this checkpoint. If a mode is missing, continue the exact same run.
state = sampler.resume_until_goal(
    state,
    evidence_goal(0.05),
    depth_cond=depth_condition,
)
state = sampler.resume_until_goal(
    state,
    evidence_goal(0.025),
    depth_cond=depth_condition,
)

results = state.to_result().trim()
```

Use targets appropriate to the problem; halving is only a convenient way to
explore. Preserve the returned `state` after every checkpoint. Resuming it
continues the same random stream, sampler geometry, samples, and allocation
history instead of paying for a restart.

## What to check at each checkpoint

- Inspect mode occupancy and posterior mass using problem-specific labels or
  regions, not only a corner plot.
- Repeat independent seeds for difficult or high-stakes problems. One run can
  show that a mode survived, but cannot establish a failure rate.
- Distinguish finding a mode from weighting it correctly. A tiny recovered
  mass may clear a “not vanished” threshold while remaining badly biased.
- Treat `log_Z_uncert` as a useful allocation goal, not as proof of multimodal
  convergence. It measures expected shrinkage uncertainty; it does not include
  uncertainty from a mode the constrained sampler failed to represent.

`unlimited_samples=True` makes this experimentation convenient because sample
storage grows geometrically and the run resumes automatically. Each new buffer
shape has a one-time compilation cost, and growth can consume unbounded memory.
Use a large finite `max_samples` instead when you need a hard resource ceiling.

In the investigation tracked in [#269](https://github.com/Joshuaalbert/jaxns/issues/269),
one persistent spike--slab failure still had a vanished weak mode at
`log_Z_uncert=0.020` after 31,840 lineages and 777,218 samples. Continuing the
same experiment to `0.016` used 49,760 lineages and 1,213,904 samples and
finally crossed the chosen mode-survival threshold. That target is not a
general recommendation; it illustrates why experimentation, resumability, and
a sufficiently large sample allowance matter.

If progressively tighter goals remain impractical or the recovered mode weight
does not stabilise, the problem likely needs many more initial lineages or a
future explicitly mode-preserving allocation strategy rather than another
small change in stopping tolerance.
